feat: schematics for guide (2) - modernize-apps-migrated-from-2211.32-to-2211.35.md

Platonn opened the pull request:
Schematics automating the guide https://github.com/SAP/spartacus/blob/develop/docs/migration/2211_35/modernize-apps-migrated-from-2211.32-to-2211.35.md

It's a part 2 of 2 https://jira.tools.sap/browse/CXSPA-9304



QA steps:
- create new ng17 app with Spa 2211.19 WITH SSR

- migrate it to ng19 and Spa 2211.35.0-1 according to this doc draft https://github.com/SAP/spartacus/blob/develop/docs/migration/2211_35/migration.md

- build and publish to verdaccio libraries from this branch

- run the migration schematics in the following way (it will be the recommended way for customers:

```bash
# Execute separate migration schematic:
ng g @spartacus/schematics:modernize-app-migrated-from-2211_32-to-2211_35
```

- check the diff of changed files in your app and compare against the changes mentioned in the manual guide: https://github.com/SAP/spartacus/blob/develop/docs/migration/2211_35/modernize-apps-migrated-from-2211.32-to-2211.35.md

- repeat everything, but starting with the app with CSR ONLY (no SSR)

Platonn committed:
initial version

Platonn committed:
make the functions for printing fallback docs to contain the version names - to not be confused by IDE with newer versions of docs in the future migrations

Platonn committed:
use aliased import of util function, for readability

Platonn committed:
add JSDocs and improve error handling

Platonn committed:
fix description of schematic in collection.json

Platonn committed:
Merge branch 'kris-schematics--guide1-develop' into kris-schematics--guide2-develop

Platonn committed:
wording of FALLBACK_ADVICE_TO_FOLLOW_DOCS + dont export it

Platonn committed:
improve logic, fix error with tree.create(directory), improve logs

github-actions[bot] committed:
Add license header

pawelfras commented on projects/schematics/src/modernize-app-migrated-from-2211_32-to-2211_35/csr-and-ssr/move-assets-to-public.ts:
```diff
@@ -0,0 +1,61 @@
+        `Error moving assets file from "${oldPath}" to "${newPath}". Error: ${error}`,
+        context
+      );
+    }
```
`major:` this logic doesn't support folders (e.g. if someone keep `i18-assets` in assets folder). Let's handle this scenario too.

Platonn replied:
Thanks. good catch again!
now it's fixed

Platonn replied:
fixed

pawelfras commented on projects/schematics/src/modernize-app-migrated-from-2211_32-to-2211_35/index.ts:
```diff
@@ -0,0 +1,48 @@
+      withFallbackDocs(updateMainTs()),
+      withFallbackDocs(updateI18nLazyLoadingConfig()),
+
+      ...(isUsingSsr(tree, context)
```
`medium:` let's consider using `checkIfSSRIsUsedWithApplicationBuilder` instead of providing new util
https://github.com/SAP/spartacus/blob/f855a213478ed090091bf21be45d6e8ba51292c4/projects/schematics/src/shared/utils/package-utils.ts#L189

Platonn replied:
done

pawelfras commented on projects/schematics/src/modernize-app-migrated-from-2211_32-to-2211_35/fallback-advice-to-follow-docs.ts:
```diff
@@ -0,0 +1,45 @@
+ */
+export const DOCS_URL_FOR_MODERNIZING_APPS_MIGRATED_FROM_2211_32_TO_2211_35 =
+  'https://help.sap.com/docs/SAP_COMMERCE_COMPOSABLE_STOREFRONT/10a8bc7f635b4e3db6f6bb7880e58a7d/525dd86b812544249e78d73f67f3e6ed.html';
+
```
`major:` This link refers to modernizing apps from 6.8 to 2211.19. If I'm not mistaken, here's the proper one: https://help.sap.com/docs/SAP_COMMERCE_COMPOSABLE_STOREFRONT/10a8bc7f635b4e3db6f6bb7880e58a7d/9bb7dd061da14cfc9c5e3f36388c77e4.html

Platonn replied:
thank you! good catch. fixed

pawelfras commented on projects/schematics/src/modernize-app-migrated-from-2211_32-to-2211_35/ssr/update-server-ts.ts:
```diff
@@ -0,0 +1,55 @@
+ */
+export function updateServerTs(): Rule {
+  return (tree: Tree, context: SchematicContext) => {
+    const oldPath = 'server.ts';
+    const newPath = 'src/server.ts';
```
`nitpick:` In PR with schematics for guide 1 is a following pattern for handling old and new paths/names:
```ts
enum AppServerModulePaths {
    old = 'src/app/app.server.module.ts',
    new = 'src/app/app.module.server.ts',
}
```

For consistency, let's consider unifying the approach (there are few more places in this PR that apply)

Platonn replied:
done

Platonn committed:
Merge branch 'kris-schematics--guide1-develop' into kris-schematics--guide2-develop

Platonn committed:
Merge branch 'kris-schematics--guide2-develop' of github.com:SAP/spartacus into kris-schematics--guide2-develop

Platonn committed:
fix URL for docs

Platonn committed:
use existing util checkIfSSRIsUsedWithApplicationBuilder

Platonn committed:
use enum for old/new path of server.ts

Platonn committed:
fix moveAssetsToPublic(). previously it only deleted files instead of moving them

Platonn committed:
make logs about moved files more concise

Platonn committed:
fix logs for moving favico

Platonn committed:
use enum for moveAssetsToPublic

Platonn committed:
fix(update i18n lazy loading config): allow for both syntax:

`loader: (lang: string, chunkName: string) => {`
and
`loader(lang: string, chunkName: string) {`

Platonn committed:
refactor: rename checkIfSSRIsUsedWithApplicationBuilder to isSsrUsed

Platonn committed:
Merge branch 'kris-schematics--guide1-develop' into kris-schematics--guide2-develop

Platonn committed:
Merge branch 'kris-schematics--guide1-develop' into kris-schematics--guide2-develop

pawelfras reviewed (approved):
review and QA done. LGTM.
Tested migration agains few different scenarios:
- app migrated from Spartacus 6.8 to 2211.19 then to 2211.35
- app with SSR migrated from Spartacus 6.8 to 2211.19 , then to 2211.35
- app migrated from 2211.19 to 2211.35
- app with SSR migrated from 2211.19 to 2211.35
- assets containing also folders
- translation with/without lazy loading

Platonn committed:
Merge branch 'kris-schematics--guide1-develop' into kris-schematics--guide2-develop

Platonn committed:
empty commit

Platonn committed:
Merge branch 'kris-schematics--guide1-develop' into kris-schematics--guide2-develop

Platonn committed:
don't use ApplicationBuilderOptions

Platonn commented:
Merged as a partial PR to another PR
